x/oracle/internal/types/msgs.go

```diff
@@ -212,3 +212,144 @@ func (msg MsgDelegateFeedConsent) String() string {
 	delegate:   %s`,
 		msg.Operator, msg.Delegate)
 }
+
+// MsgAggregateExchangeRatePrevote - struct for aggregate prevoting on the ExchangeRateVote.
+// The purpose of aggregate prevote is to hide vote exchange rates with hash
+// which is formatted as hex string in SHA256("{salt}:{exchange rate}{denom},...,{exchange rate}{denom}:{voter}")
+type MsgAggregateExchangeRatePrevote struct {
+	Hash      AggregateVoteHash `json:"hash" yaml:"hash"`
+	Feeder    sdk.AccAddress    `json:"feeder" yaml:"feeder"`
+	Validator sdk.ValAddress    `json:"validator" yaml:"validator"`
+}
+
+// NewMsgAggregateExchangeRatePrevote returns MsgAggregateExchangeRatePrevote instance
+func NewMsgAggregateExchangeRatePrevote(hash AggregateVoteHash, feeder sdk.AccAddress, validator sdk.ValAddress) MsgAggregateExchangeRatePrevote {
+	return MsgAggregateExchangeRatePrevote{
+		Hash:      hash,
+		Feeder:    feeder,
+		Validator: validator,
+	}
+}
+
+// Route implements sdk.Msg
+func (msg MsgAggregateExchangeRatePrevote) Route() string { return RouterKey }
+
+// Type implements sdk.Msg
+func (msg MsgAggregateExchangeRatePrevote) Type() string { return "aggregateexchangerateprevote" }
+
+// GetSignBytes implements sdk.Msg
+func (msg MsgAggregateExchangeRatePrevote) GetSignBytes() []byte {
+	return sdk.MustSortJSON(ModuleCdc.MustMarshalJSON(msg))
+}
+
+// GetSigners implements sdk.Msg
+func (msg MsgAggregateExchangeRatePrevote) GetSigners() []sdk.AccAddress {
+	return []sdk.AccAddress{msg.Feeder}
+}
+
+// ValidateBasic Implements sdk.Msg
+func (msg MsgAggregateExchangeRatePrevote) ValidateBasic() sdk.Error {
+
```

Comment:
Shouldn't you validate the exchangeRates string here? e.g. check for correct comma separation, etc

Reply:
In prevote, we do not receive exchangeRates string. We are doing this in vote ValidateBasic() with `ParseExchangeRateTuples` function

Reply:
Makes sense